journalbeat/reader/journal.go

```diff
@@ -179,24 +181,80 @@ func (r *Reader) seek(cursor string) {
 // Next waits until a new event shows up and returns it.
 // It blocks until an event is returned or an error occurs.
 func (r *Reader) Next() (*beat.Event, error) {
+	event, err := r.readEvent()
+	if err != nil {
+		return nil, err
+	}
+	if event != nil {
+		return event, nil
+	}
+
+	return r.waitUntilNewEventOrError()
+}
+
+func (r *Reader) waitUntilNewEventOrError() (*beat.Event, error) {
+	var wg sync.WaitGroup
+	defer wg.Wait()
+
+	changesChan := make(chan int)
+	defer close(changesChan)
+
 	for {
 		select {
 		case <-r.done:
 			return nil, nil
 		default:
-			event, err := r.readEvent()
-			if err != nil {
-				return nil, err
+			for {
+				wg.Add(1)
+				go r.waitForChange(&wg, changesChan)
+
+				select {
+				case <-r.done:
+					return nil, nil
+				case c, ok := <-changesChan:
+					if !ok {
+						break
```

Comment:
ok is `false` if `changesChan` is closed. The worker is not closing it, so there is no need to check for the bool. Yet the `break` statement targets the inner `select`, such that we continue with the inner for-loop.

Let's try to reduce indentation (why do we need the inner loop?):

```
for {
  // check if shutdown signal is given
  select {
    case <- r.done:
      return nil, nil
    default:
      // nothing
  }

  wg.Add(1)
  go r.waitForChange(&wg, changesChan)

  var c int
  select {
  case <-r.done:
    // shutdown signal
    return nil, nil
  case c = <-changesChan:
  }
    
  switch c {
      ...
  }
}
```

Starting a go-routine per event sounds quite expensive. How about starting the second go-routine at the beginning and signalling it (via a channel) that it needs to do work.